File: [0295][Hard][Find_Median_from_Data_Stream][Heap_Design]/Find_Median_from_Data_Stream_2.py
# leetcode time     cost : 244 ms
# leetcode memory   cost : 24.5 MB 
# solution 2, insert sorting
# Time  Complexity: O(NlogN)
# Space Complexity: O(1)
import bisect
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """

        self.data = []

    def addNum(self, num: int) -> None:
        bisect.insort_left(self.data, num)

    def findMedian(self) -> float:
        n = len(self.data)
        mid = (n - 1) // 2
        if n % 2 == 1:
            return self.data[mid]
        else:
            return (self.data[mid] + self.data[mid + 1]) / 2


def main():
    obj = MedianFinder()
    obj.addNum(1)
    obj.addNum(2)
    obj.addNum(3)
    param_2 = obj.findMedian()
    print("in python,res is : ",param_2)
if __name__ =='__main__':
    main() 

File: [0295][Hard][Find_Median_from_Data_Stream][Heap_Design]/Find_Median_from_Data_Stream_2.cpp
// solution 2, insert sorting
// leetcode time     cost : 452 ms
// leetcode memory   cost : 41.9 MB
// Time  Complexity: O(N+logN)=O(N)
// Space Complexity: O(N)
#include <string.h>
#include <stdlib.h>
#include <iostream>
#include <queue>
#include <algorithm>
using namespace std;

class MedianFinder {
    vector<int> store; // resize-able container

public:
    // Adds a number into the data structure.
    void addNum(int num)
    {
        if (store.empty())
            store.push_back(num);
        else
            // binary search and insertion combined
            store.insert(lower_bound(store.begin(), store.end(), num), num);     
    }

    // Returns the median of current data stream
    double findMedian()
    {
        int n = store.size();
        return n & 1 ? store[n / 2] : (store[n / 2 - 1] + store[n / 2]) * 0.5;
    }
};

int main(){
    MedianFinder* obj = new MedianFinder();
    obj->addNum(1);
    obj->addNum(2);
    obj->addNum(3);
    double param_2 = obj->findMedian();
    cout<<param_2;
    return 0;
}